pkg/cmd/pulumi/templates/templates.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2016-2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package templates adds an abstraction for project templates that may be local or
+// remote.
+//
+// All templates are convertible into [workspace.Template].
+package templates
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+// Source provides access to a set of project templates, any set of which may be present on
+// disk.
+//
+// Source is responsible for cleaning up old templates, and should always be [Close]d when
+// created.
+type Source struct {
+	templates    []Template
+	errorOnEmpty []error
+	errors       []error
+	// closers holds a list of functions to be invoked when the Source is closed.
+	//
+	// closers always has length at least one, and the first closer always cancels the
+	// context associated with fetch operations.
+	closers []func() error
+	closed  bool
+
+	// m should be held whenever Source is mutated.
+	m  sync.Mutex
+	wg sync.WaitGroup
+}
+
+// Templates lists the templates available to the [Source].
+//
+// Templates *does not* produce a sorted list. If templates need to be sorted, then the
+// caller is responsible for sorting them.
+func (s *Source) Templates() ([]Template, error) {
+	s.wg.Wait() // Wait to ensure that all templates have been fetched before returning the template list.
+
+	s.lockOpen("read templates")
+	defer s.m.Unlock()
+	if err := errors.Join(s.errors...); err != nil {
+		return nil, err
+	}
+	if len(s.templates) == 0 {
+		return nil, errors.Join(s.errorOnEmpty...)
+	}
+	return s.templates, nil
+}
+
+func (s *Source) addTemplate(t Template) {
+	contract.Assertf(t != nil, "We should never return nil templates")
+	s.lockOpen("add template")
+	s.templates = append(s.templates, t)
+	s.m.Unlock()
+}
+
+func (s *Source) addCloser(f func() error) {
+	s.lockOpen("add closer")
+	s.closers = append(s.closers, f)
```

Comment:
CI seems to detect a race here.  Not entirely sure how that's possible since we do seem to have a lock used, but we should look at that.

Reply:
I bet it's actually related to https://github.com/pulumi/pulumi/pull/18627#discussion_r2009911704. `(*Source).addCloser` can be called at the same time as `(*Source).Close`, and `(*Source).Close` used to call `s.closers[0]()` **before**[^1] it takes the lock.

[^1]: This was necessary since canceled functions are still allowed to add closers, but also potentially racy if `s.closers` was re-allocating while cancel was being called.